lib/nosedrum/storage.ex

```diff
@@ -1,75 +1,142 @@
 
-  The public-facing API of storage modules should use this in order
-  to allow users to identify the command they want to operate on.
+  **Note** that Discord only supports nesting 3 levels deep, like `command -> subcommand group -> subcommand`.
 
-  ## Usage
-  To identify a single command, use a single element list, such as `["echo"]`.
-  To identify a subcommand, use a pair, such as `["infraction", "search"]`.
-  To identify the default subcommand invoked when no matching subcommand is
-  found, specify the group name first, then `:default`, such as
-  `["tags", :default]`.
+  ## Example path:
+  ```elixir
+  %{
+    {"castle", MyApp.CastleCommand.description()} =>
+      %{
+        {"prepare", "Prepare the castle for an attack."} => [],
+        {"open", "Open up the castle for traders and visitors."} => [],
+        # ...
+      }
+  }
+  ```
+  """
+  @type application_command_path ::
+          %{
+            {group_name :: String.t(), group_desc :: String.t()} => [
+              application_command_path | [Nosedrum.ApplicationCommand.option()]
+            ]
+          }
+
+  @typedoc """
+  The name or pid of the Storage process.
   """
-  @type command_path :: [String.t() | :default, ...]
+  @type name_or_pid :: atom() | pid()
 
   @doc """
-  Look up a command group under the specified `name`.
+  Handle an Application Command invocation.
 
-  If the command was not found, `nil` should be returned.
+  This callback should be invoked upon receiving an interaction via the `:INTERACTION_CREATE` event.
+
+  ## Example using `Nosedrum.Storage.Dispatcher`:
+  ```elixir
+  # In your `Nostrum.Consumer` file:
+  def handle_event({:INTERACTION_CREATE, interaction, _ws_state}) do
+    IO.puts "Got interaction"
+    Nosedrum.Storage.Dispatcher.handle_interaction(interaction)
+  end
+  ```
   """
-  @callback lookup_command(name :: String.t(), storage :: reference) :: command_group | nil
+  @callback handle_interaction(interaction :: Interaction.t(), name_or_pid) ::
+              {:ok} | {:error, :unknown_command} | Nostrum.Api.error()
 
   @doc """
-  Add a new command under the given `path`.
+  Add a new command under the given name or application command path. Returns `:ok` if successful, and
+  `{:error, reason}` otherwise.
 
-  If a command has the c:Nosedrum.Command.aliases/0 callback defined,
-  they will also be added under `path`. If the command already exists,
-  no error should be returned.
+  If the command already exists, it will be overwritten.
   """
-  @callback add_command(path :: command_path, command :: module, storage :: reference) ::
-              :ok | {:error, String.t()}
+  @callback add_command(
+              name_or_path :: String.t() | application_command_path,
+              command_module :: module,
+              scope :: command_scope,
+              name_or_pid
+            ) ::
+              :ok | {:error, reason :: String.t()}
```

Comment:
Do the error cases really return strings? It looks like they return the nostrum error here.

Reply:
Same for `remove_command`

Reply:
Oh yes it definitely shouldn't be a String exclusively since Dispatcher return `ApiError`s, but I now wonder if we should even have concrete types here for these callbacks. For example, if a user implements this behaviour, maybe they want to reject adding/removing a command for reasons other than an API error? What do you think?

Reply:
In case we don't want a weaker/more flexible return value as per above, I changed `String.t()` to `Nostrum.Error.ApiError.t()` for now

Reply:
> Oh yes it definitely shouldn't be a String exclusively since Dispatcher return ApiErrors, but I now wonder if we should even have concrete types here for these callbacks. For example, if a user implements this behaviour, maybe they want to reject adding/removing a command for reasons other than an API error? What do you think?

I think we can stick to changing it to ApiError for now like you did, and if somebody has a need for it add support for it. That said I think these behaviours can also be implemented on top of it, so 🤷